### PR TITLE
KAFKA-17570: Rewrite StressTestLog by Java

### PR DESCRIPTION
**Pull request by frankvicky**
JIRA: [KAFKA-17570](https://issues.apache.org/jira/browse/KAFKA-17570)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: jmh-benchmarks/src/main/java/org/apache/kafka/jmh/log/StressTestLog.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        reader.start();
+        while (running.get()) {
+            Thread.sleep(1000);
+            System.out.printf("Reader offset = %d, writer offset = %d%n", reader.getCurrentOffset(), writer.getCurrentOffset());
+            writer.checkProgress();
+            reader.checkProgress();
+        }
+    }
+
+    abstract class WorkerThread extends Thread {
```
Comment (chia7712) [inline jmh-benchmarks/src/main/java/org/apache/kafka/jmh/log/StressTestLog.java]: Could you please use `@Threads` instead

File: jmh-benchmarks/src/main/java/org/apache/kafka/jmh/log/StressTestLog.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+                    VerificationGuard.SENTINEL);
+            assert (logAppendInfo.firstOffset() == -1 || logAppendInfo.firstOffset() == currentOffset) && logAppendInfo.lastOffset() == currentOffset;
+            currentOffset++;
+            if (currentOffset % 1000 == 0) {
+                Thread.sleep(50);
+            }
+        }
+    }
+
+    class ReaderThread extends WorkerThread {
```
Comment (chia7712) [inline jmh-benchmarks/src/main/java/org/apache/kafka/jmh/log/StressTestLog.java]: Could you please rewrite it by `@Threads`?

File: jmh-benchmarks/src/main/java/org/apache/kafka/jmh/log/StressTestLog.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+        ReaderThread(UnifiedLog log) {
+            super(log);
+        }
+
+        @Override
+        public void work() {
+            Records read = log.read(currentOffset, 1, FetchIsolation.LOG_END, true).records;
```
Comment (chia7712) [inline jmh-benchmarks/src/main/java/org/apache/kafka/jmh/log/StressTestLog.java]: in this test case, we should prepare a log with data before testing read.

**Review by chia7712 — COMMENTED** [review]
@frankvicky Except for read/write test, could you please test write + read mixed?

**Comment by chia7712** [comment]
@frankvicky  Could you please rewrite it by java like #17246?

**Comment by chia7712** [comment]
@frankvicky could you please attach the result ?

**Comment by frankvicky** [comment]
Hi @chia7712 
I have let it run for a while, here is the result:
```
Reader offset = 14000, writer offset = 14000
Reader offset = 26853, writer offset = 29000
Reader offset = 42000, writer offset = 42000
Reader offset = 59748, writer offset = 60000
Reader offset = 77000, writer offset = 77000
Reader offset = 94569, writer offset = 95000
Reader offset = 112008, writer offset = 112123
Reader offset = 130000, writer offset = 130000
Reader offset = 147707, writer offset = 148000
Reader offset = 165212, writer offset = 166000
Reader offset = 183000, writer offset = 183000
Reader offset = 200749, writer offset = 201000
Reader offset = 218000, writer offset = 218000
Reader offset = 235484, writer offset = 236000
Reader offset = 253000, writer offset = 253000
Reader offset = 271000, writer offset = 271000
Reader offset = 287721, writer offset = 288000
Reader offset = 304000, writer offset = 304005
Reader offset = 319234, writer offset = 320000
Reader offset = 336628, writer offset = 337000
Reader offset = 353424, writer offset = 354000
Reader offset = 370000, writer offset = 370000
Reader offset = 387053, writer offset = 387855
Reader offset = 404005, writer offset = 404129
Reader offset = 421521, writer offset = 422000
Reader offset = 439057, writer offset = 440000
Reader offset = 456800, writer offset = 457000
Reader offset = 472000, writer offset = 472000
Reader offset = 489473, writer offset = 490000
Reader offset = 507000, writer offset = 507000
Reader offset = 524000, writer offset = 524000
Reader offset = 541160, writer offset = 542000
Reader offset = 558988, writer offset = 559000
Reader offset = 576000, writer offset = 576000
Reader offset = 593682, writer offset = 594000
Reader offset = 611000, writer offset = 611000
Reader offset = 628000, writer offset = 628000
Reader offset = 645873, writer offset = 646000
Reader offset = 663062, writer offset = 664000
Reader offset = 680976, writer offset = 681000
Reader offset = 698000, writer offset = 698000
Reader offset = 716000, writer offset = 716000
Reader offset = 733000, writer offset = 733000
Reader offset = 750000, writer offset = 750000
Reader offset = 768000, writer offset = 768000
Reader offset = 785029, writer offset = 785520
Reader offset = 802484, writer offset = 803000
Reader offset = 820000, writer offset = 820000
Reader offset = 837000, writer offset = 837000
Reader offset = 854000, writer offset = 854000
Reader offset = 871228, writer offset = 872000
Reader offset = 888000, writer offset = 888000
Reader offset = 899547, writer offset = 902000
Reader offset = 917766, writer offset = 918000
Reader offset = 935000, writer offset = 935000
Reader offset = 953000, writer offset = 953000
Reader offset = 970000, writer offset = 970000
Reader offset = 987614, writer offset = 988000
Reader offset = 1005273, writer offset = 1006000
Reader offset = 1022153, writer offset = 1023000
Reader offset = 1039464, writer offset = 1040000
Reader offset = 1057000, writer offset = 1057000
Reader offset = 1074382, writer offset = 1075000
Reader offset = 1092000, writer offset = 1092000
Reader offset = 1110000, writer offset = 1110000
Reader offset = 1126000, writer offset = 1126000
Reader offset = 1143808, writer offset = 1144000
Reader offset = 1161000, writer offset = 1161000
Reader offset = 1177626, writer offset = 1178508
Reader offset = 1195133, writer offset = 1196000
Reader offset = 1213000, writer offset = 1213000
Reader offset = 1230743, writer offset = 1231000
Reader offset = 1248000, writer offset = 1248000
Reader offset = 1266000, writer offset = 1266000
Reader offset = 1283463, writer offset = 1284000
Reader offset = 1301000, writer offset = 1301000
Reader offset = 1318655, writer offset = 1319000
Reader offset = 1336535, writer offset = 1337000
Reader offset = 1353000, writer offset = 1353000
Reader offset = 1371000, writer offset = 1371000
Reader offset = 1388315, writer offset = 1389000
Reader offset = 1402628, writer offset = 1405000
Reader offset = 1413757, writer offset = 1418046
```

**Review by chia7712 — APPROVED** [review]

**Pull request merged**